packages/adelus/src/factor.hpp

```diff
@@ -0,0 +1,1130 @@
+
+  double pivot_mag;
+  value_type invpiv; //KK
+
+  //DATA_TYPE *cur_col_ptr,*cur_row_ptr,*update_ptr,*piv_row_ptr;
+  //DATA_TYPE *sav_col_ptr,*sav_row_ptr,*sav_piv_row_ptr;
+  //DATA_TYPE *cur_col1_row_ptr,*piv_col1_row_ptr;
+  //DATA_TYPE *temp_row_ptr;
+  //DATA_TYPE *act_col_ptr,*act_row_ptr,*act_piv_row_ptr;
+  ////DATA_TYPE *act_cur_row_ptr;//Note: comment out since act_cur_row_ptr is not used
+
+  int gpivot_row; /* make sure that this is well aligned */
+
+  int cur_col_i, cur_col_j, cur_row_i, cur_row_j, act_col_i, act_row_j, update_i, update_j; //KK
+  int sav_col_i, sav_col_j, sav_piv_row_i, sav_piv_row_j, act_piv_row_i, piv_row_i; //KK
+  int cur_col1_row_i, piv_col1_row_i; //KK
+
+  int /*dest,*/ringdist,rdist;
+  long type,bytes;
+
+
+#ifdef DREAL
+  struct {
+    double  val;
+    int proc;
+  } pivot_in,pivot_out;
+#endif
+#ifdef ZCPLX
+  struct {
+    double  val;
+    int proc;
+  } pivot_in,pivot_out;
+#endif
+#ifdef SREAL
+  struct {
+    float  val;
+    int proc;
+  } pivot_in,pivot_out;
+#endif
+#ifdef SCPLX
+  struct {
+    float  val;
+    int proc;
+  } pivot_in,pivot_out;
+#endif
+
+  //DATA_TYPE entry,current;
+  value_type entry,current;
+  int row;
+
+  int numprocs;
+
+#ifdef GET_TIMING
+  double updatetime,colupdtime,rowupdtime,scaltime;
+  double xpivmsgtime,bcastpivstime,bcastpivrtime,bcastcolstime,bcastcolrtime,bcastrowtime,sendrowtime,recvrowtime;
+  double copycoltime,copyrowtime,copyrow1time,copypivrowtime,copypivrow1time,pivotswaptime;
+  double t1,t2;
+  double msgtime,copytime,dgemmtime,totalfactortime;
+  double iamaxtime,getlocalpivtime,localpivtime;
+#if defined(CUDA_HOST_PINNED_MPI) && defined(KOKKOS_ENABLE_CUDA)
+  double copyhostpinnedtime;
+#endif
+#endif
+
+  MPI_Request msgrequest;
+  MPI_Status msgstatus;
+  /* Distribution for the matrix on me */
+
+  MPI_Comm_size(MPI_COMM_WORLD,&numprocs);
+  if ( (numprocs/nprocs_row) * nprocs_row != numprocs )
+  {
+     if (me == 0)
+     {
+       printf("nprocs_row must go into numprocs perfectly!\n");
+       printf("Try a different value of nprocs_row.\n");
+     }
+     MPI_Barrier(MPI_COMM_WORLD);
+     exit(0);
+  }
+
+  //DATA_TYPE *seg  = reinterpret_cast<DATA_TYPE *>(ZV.data());
+  //DATA_TYPE *col1 = reinterpret_cast<DATA_TYPE *>(col1_view.data());  /* ptrs to col used for updating a col */
+  //DATA_TYPE *row1 = reinterpret_cast<DATA_TYPE *>(row1_view.data());  /* ptr to diagonal row */
+  //DATA_TYPE *row2 = reinterpret_cast<DATA_TYPE *>(row2_view.data());  /* ptr to pivot row */
+  //DATA_TYPE *row3 = reinterpret_cast<DATA_TYPE *>(row3_view.data());  /* ptr to temporary vector for rows */
+  //int *pivot_vec  = reinterpret_cast<int *>(pivot_vec_view.data());   /* ptr to vector storing list of pivot rows */
+
+  printf("Rank %i -- factor() Begin LU factorization, execution_space %s, memory_space %s\n", me, typeid(execution_space).name(), typeid(memory_space).name());
```

Comment:
I am typically hesitant to recommend output like this in a library - especially if it cannot be disabled.  Output is a fairly slow operation on most systems.  Consider carefully if you really want the printf statements that are embedded throughout.  If so, you may consider finding a way to make them optional in case a downstream library doesn't want them.

Reply:
I made printf optional. Users can enable printf in Adelus when building Trilinos.